test/annexB/language/statements/do-while/labeled-fn-stmt.js

```diff
@@ -0,0 +1,10 @@
+// Copyright (C) 2016 the V8 project authors. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+description: It is a Syntax Error if IsLabelledFunction(Statement) is true.
+negative: SyntaxError
+id: sec-labelled-function-declarations
+es6id: B.3.2
+---*/
+
+do label1: label2: function f() {} while (false)
```

Comment:
This is a little tricky, but I believe this particular early error applies
regardless of language mode.

[Section
13.3.1](http://www.ecma-international.org/ecma-262/6.0/#sec-labelled-statements-static-semantics-early-errors)
describes an early error for labelled statements generally:

> LabelledItem : FunctionDeclaration
>
>     It is a Syntax Error if any source text matches this rule.
> 
> NOTE An alternative definition for this rule is provided in B.3.2.

[Section
B.3.2](http://www.ecma-international.org/ecma-262/6.0/#sec-labelled-function-declarations)
modifies that general rule:

> It is a Syntax Error if any strict mode source code matches this rule.

This clearly limits the error to strict mode. But the early error under test
specifically concerns the statement as is appears within an IterationStatement.
[Section
13.7.1.1](http://www.ecma-international.org/ecma-262/6.0/#sec-labelled-statements-static-semantics-early-errors)
reads:

> It is a Syntax Error if IsLabelledFunction(Statement) is true.
>
> NOTE It is only necessary to apply this rule if the extension specified in
> B.3.2 is implemented.

Here, the language mode is *not* taken into account.

So by my reading, the only place where a labelled function declaration should
not throw an error is when it appears in a StatementList of non-strict mode
code that does not implement Annex B:

Position           | Mode       | Annex B         | Error?
-------------------|------------|-----------------|--------
StatementList      | Non-strict | Not implemented | Yes
StatementList      | Non-strict | Implemented     | No
StatementList      | Strict     | Not implemented | Yes
StatementList      | Strict     | Implemented     | Yes
IterationStatement | Non-strict | Not implemented | Yes
IterationStatement | Non-strict | Implemented     | Yes
IterationStatement | Strict     | Not implemented | Yes
IterationStatement | Strict     | Implemented     | Yes

Does this sound right to you?